### PR TITLE
circle_tool_v2 project

### LINKED ISSUE #1618: taigaclient: update models
models are defined here https://github.com/threefoldtech/circles_reporting_tool/blob/master/specs/models.md, original implementation https://github.com/threefoldtech/js-sdk/blob/357f5aeb810a0dc463f8b96042e0b548862aaffa/jumpscale/clients/taiga/models.py#L379

### LINKED ISSUE #1619: taigaclient: update circle creation
more specs added  for projects, team, funnels covering statuses for stories, tasks and issues and custom fields

- [x] team https://github.com/threefoldtech/circles_reporting_tool/blob/master/specs/team.md
- [x] project https://github.com/threefoldtech/circles_reporting_tool/blob/master/specs/project.md
- [x] funnel  https://github.com/threefoldtech/circles_reporting_tool/blob/master/specs/funnel.md
  - [x] funnel custom fields https://github.com/threefoldtech/circles_reporting_tool/blob/master/specs/funnel.md#custom-fields

### LINKED ISSUE #1620: taigaclient: update export
Right now export exports projects, user stories, issues, users, milestones let's make sure it conforms it works as expected and also  to include tasks objects as well

### LINKED ISSUE #1621: taigaclient: import feature
Allow importing from export dir

https://github.com/threefoldtech/circles_reporting_tool/blob/master/specs/export_import.md

### LINKED ISSUE #1622: taigaclient: archive circle type
Add new type archived circle and don't use it in exports https://github.com/threefoldtech/circles_reporting_tool/blob/master/specs/reportingtool.md#archive

### LINKED ISSUE #1623: taigaclient: pull updates every 5 mins
In taigaclient define a command to export as md every 5 mins

### LINKED ISSUE #1660: Corrupted exported yaml files in taiga client
### Description
Found a corrupted format in the exported YAML files in taiga client

### PR DESCRIPTION
**Pull request by MElborolossy**
### Description

- Update models, get and validate custom fields in funnel circle as described in [funnel custom fields](https://github.com/threefoldtech/circles_reporting_tool/blob/master/specs/funnel.md#custom-fields).
- Check export work as expected.
- Create ArchiveCircle Model and exclude it from export.
- Fix export as YAML and include all the object attributes in the exported files.
- Create export every 5 min and add an option to update the modified objects only.
- Import from Yaml "Some Values will be replaced with default one".
- Enhance Cross linking in the wiki _"md_files"_.
- Wiki can be used with publishingtools.
- Update documentations.

### Changes

- Update models with the latest specs.
- Create get_issue_custom_fields function.
- Create get_user_story_custom_fields functions.
- Create validate_custom_fields function.
- Add custom field [bookings, commission] in create_funnel_circle.
- Add custom_fields as attribute in _create_new_circle function.
- Update __dir__ in models.py with missed attributes and methods.
- Create CircleTask Class.
- Add tasks in missed places in models.py file.
- Export as YAML include tasks now.
- Create ArchiveCircle Class.
- Update export and as_md functions to exclude ArchiveCircle.
- Create export periodically of time.
- Add an option to update the modified objects only.
- Import from YAML files.
- Update taiga_client tutorial and taiga docstrings with the new features.

### Related Issues
- [x] Resolves #1618 -- Commit d2a544a2fba56ddedfef7b87f7b36ee82c361479
- [x] Resolves #1619 -- Commit d2a544a2fba56ddedfef7b87f7b36ee82c361479
- [x] Resolves #1620 -- Commit 1dc2375d8e77e11f1dbb6caadf12a9a6b3feb145
- [x] Resolves #1621 -- Commit 8baddd6b66c1ed092b245235e0587971a5fe31e7
- [x] Resolves #1622 -- Commit 2769bf8f03feb1926cae82db44251a8774b6c8bc
- [x] Resolves #1623 -- Commit 6497d8c45c0c78636430eff39baa822fdddd0042
- [x] Resolves #1660 -- Commit 41b520473e557d49282715b1b9e5f2a26722bada

### Checklist
- [x] Build pass
- [x] Documentation

### CODE REVIEW

**Comment by MElborolossy** [comment]
For Export and Import, I found if we use JSON, it will be better and easier than YAML, so that we can export and import using the Rest API as [Export and Import Taiga REST API](http://taigaio.github.io/taiga-doc/dist/api.html#export-import-export-dump)

**Review by xmonader — APPROVED** [review]

**Review by xmonader — CHANGES_REQUESTED** [review]
Added some comments and let's try using gevent.event instead of threading module

File: jumpscale/clients/taiga/models.py [hunk 1]
```diff
@@ -9,11 +9,25 @@ def __init__(self, taigaclient, original_object):
         self._original_object = original_object
         self._client = taigaclient
         self._api = self._client.api
+    
+    def add_allowed_params(self, *args):
```
[the remark this answers is not in the document]
Comment (MElborolossy) [reply jumpscale/clients/taiga/models.py:13]: **allowed_params** is a list in the Taiga client that contains the parameters that will be represented as dict in to_dict() function, as we use to_dict() to export the objects, so we need to add more parameter to the same list for each model type.

File: jumpscale/clients/taiga/taiga.py [hunk 2]
```diff
@@ -537,6 +604,79 @@ def list_project_circles(self):
...
+
+            if confidence % 10 != 0:
+                j.exceptions.Validation("Confidence: ({}) not multiple of 10, it must be multiple of 10".format(confidence))
+            
+            part_tmp = part.replace('%', '')
+            if user != None and user not in self.list_all_users():
+                raise j.exceptions.Validation('User: ({}) is not found'.format(user))
+
+            if  int(part_tmp) < 0 or int(part_tmp) > 100:
+                j.exceptions.Validation('Part: ({}) is a not a valid percentage, it must be from 0% to 100%')
```
[the remark this answers is not in the document]
Comment (MElborolossy) [reply jumpscale/clients/taiga/taiga.py:672]: I will fix it the next commit

File: jumpscale/clients/taiga/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MElborolossy) [reply jumpscale/clients/taiga/models.py]: BTW I will build the YAML file without using this method.

File: jumpscale/clients/taiga/models.py [hunk 3]
```diff
@@ -66,23 +94,80 @@ def as_md(self):
...
+    @property
+    def url(self):
+        # https://circles.threefold.me/project/despiegk-tftech-software/task/286
+        return f"{self._client.host}/project/{self.project_extra_info.get('slug')}/task/{self.id}"
 
+    
 
+    @property
+    def as_md(self):
+        TEMPLATE = dedent(
```
[the remark this answers is not in the document]
Comment (MElborolossy) [reply jumpscale/clients/taiga/models.py:155]: Fix cross-linking, it is not using id.
it is using ref instead.

**Review by xmonader — APPROVED** [review]

**Pull request merged**